chia/util/struct_stream.py

```diff
@@ -13,48 +11,88 @@ def __trunc__(self) -> int:
+    if remainder != 0:
+        raise ValueError(f"cls.BITS must be a multiple of 8: {cls.BITS}")
+
+    cls.PACK = packing_strings[(cls.SIZE, cls.SIGNED)]
+
+    if cls.SIGNED:
+        cls.MAXIMUM_EXCLUSIVE = 2 ** (cls.BITS - 1)
+        cls.MINIMUM = -(2 ** (cls.BITS - 1))
+    else:
+        cls.MAXIMUM_EXCLUSIVE = 2 ** cls.BITS
+        cls.MINIMUM = 0
+
+    return cls
+
+
 class StructStream(int):
-    PACK = ""
+    PACK: Optional[str] = None
+    SIZE = 0
+    BITS = 0
+    SIGNED = False
+    MAXIMUM_EXCLUSIVE = 0
+    MINIMUM = 0
 
     """
     Create a class that can parse and stream itself based on a struct.pack template string.
     """
 
     # This is just a partial exposure of the underlying int constructor.  Liskov...
     # https://github.com/python/typeshed/blob/5d07ebc864577c04366fcc46b84479dbec033921/stdlib/builtins.pyi#L181-L185
-    def __new__(
-        cls: Type[_T_StructStream], value: Union[str, bytes, SupportsInt, SupportsIndex, SupportsTrunc]
-    ) -> _T_StructStream:
-        value = int(value)
-        try:
-            v1 = struct.unpack(cls.PACK, struct.pack(cls.PACK, value))[0]
-            if value != v1:
-                raise ValueError(f"Value {value} does not fit into {cls.__name__}")
-        except Exception:
-            bits = struct.calcsize(cls.PACK) * 8
-            raise ValueError(
-                f"Value {value} of size {value.bit_length()} does not fit into " f"{cls.__name__} of size {bits}"
-            )
-        return int.__new__(cls, value)
+    def __init__(self, value: Union[str, bytes, SupportsInt, SupportsIndex, SupportsTrunc]) -> None:
+        super().__init__()
+        if not (self.MINIMUM <= self < self.MAXIMUM_EXCLUSIVE):
+            raise ValueError(f"Value {self} does not fit into {type(self).__name__}")
 
     @classmethod
     def parse(cls: Any, f: BinaryIO) -> Any:
-        bytes_to_read = struct.calcsize(cls.PACK)
+        bytes_to_read = cls.SIZE
         read_bytes = f.read(bytes_to_read)
-        assert read_bytes is not None and len(read_bytes) == bytes_to_read
-        return cls(*struct.unpack(cls.PACK, read_bytes))
+        return cls.from_bytes(read_bytes)
 
     def stream(self, f):
-        f.write(struct.pack(self.PACK, self))
+        f.write(self.to_bytes())
 
     @classmethod
-    def from_bytes(cls: Any, blob: bytes) -> Any:  # type: ignore
-        f = io.BytesIO(blob)
-        result = cls.parse(f)
-        assert f.read() == b""
-        return result
+    def from_bytes(cls: Type[_T_StructStream], blob: bytes) -> _T_StructStream:  # type: ignore[override]
+        if len(blob) != cls.SIZE:
+            raise ValueError(f"{cls.__name__}.from_bytes() requires {cls.SIZE} bytes but got: {len(blob)}")
+        return cls(int.from_bytes(blob, "big", signed=cls.SIGNED))
+
+    def to_bytes(  # type: ignore[override]
+        self, length: Optional[int] = None, byteorder: Literal["little", "big"] = "big", signed: bool = None
```

Comment:
is `byteorder` a new feature? It would seem reasonable to me to hard code it to always be "big". Also, I don't understand why `signed` is a parameter here, shouldn't it always be `self.SIGNED`?

Reply:
I exactly agreed.  I used this as a boundary for limiting expansion of this PR, but that decision can be reconsidered.  Previously `StructStream.to_bytes()` was just inherited from `int.to_bytes()` which of course provides the general control over the serialization such as `byteorder` and `length` and `signedness`.  I'm tempted to just put the `super().to_bytes()` call into `.__bytes__()` instead and leave this definition out so it is unchanged.  Not because the code calling this is obviously good, but just as a boundary for this PR, still.

Anyways, we are calling `.to_bytes()` in a few places.

(length and byte order)
https://github.com/Chia-Network/chia-blockchain/blob/8486311afe4858d5645653d4cbd905c8fcded000/chia/util/streamable.py#L204-L205

(length and byte order)
https://github.com/Chia-Network/chia-blockchain/blob/8486311afe4858d5645653d4cbd905c8fcded000/chia/consensus/coinbase.py#L13-L18

(length and byte order and signedness)
https://github.com/Chia-Network/clvm/blob/fc73cd9dc2fc30a1fd461d0f05af9f9679e042c8/clvm/casts.py#L8-L17
```python
def int_to_bytes(v):
    byte_count = (v.bit_length() + 8) >> 3
    if v == 0:
        return b""
    r = v.to_bytes(byte_count, "big", signed=True)
    # make sure the string returned is minimal
    # ie. no leading 00 or ff bytes that are unnecessary
    while len(r) > 1 and r[0] == (0xFF if r[1] & 0x80 else 0):
        r = r[1:]
    return r
```

`Streamable.write_uint32()` and `Streamable.parse_uint32()` seem to be kind of pointless now.  I'm guessing they had some use historically when maybe we didn't have `StructStream` or something.

If you want me to expand this PR down these paths I can take a look at them.

Reply:
I think I'm confused because I don't see the definition of this `to_bytes()` on the left hand side. Would it make sense, for now, to assert that the parameters that are passed in here matches our own?